File: Merge_Sort/MergeSort.cpp
#include <iostream>
using namespace std;

void merge(int arr[] , int beg , int mid , int end)
{
    int nl = mid - beg +1 , nr = end-mid;
    int leftArr[nl] , rightArr[nr];
    
    for (int i = 0; i < nl; i++) {
        leftArr[i] = arr[beg+i];
    }
   for (int i = 0; i < nr; i++) {
        rightArr[i] = arr[mid +1 + i];
    }
    
     int i=0 , j=0 , k=beg;
    while(i<nl && j<nr){
        
        if(leftArr[i] <= rightArr[j])
        {
            arr[k]=leftArr[i];
            i++;
            k++;
        }
        else{
            arr[k]=rightArr[j];
            j++;
            k++;
        }
    }
    
    while(i<nl){
        arr[k]=leftArr[i];
            i++;
            k++;
    }
    while(j<nr){
        arr[k]=rightArr[j];
            j++;
            k++;
    }
    
}

void mergeSort(int arr[] , int beg , int end )
{
    if(beg<end)
    {
        int mid = (beg + end) /2;
        mergeSort(arr,beg,mid);
        mergeSort(arr,mid+1,end);
        merge(arr,beg,mid,end);
    }
}

void print(int arr[],int n)
{
    
    for (int i = 0; i < n; i++) {
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}
int main()
{
    int n ;
    cout << "Enter the number of elements: ";
    cin>>n;
    int * arr = new int [n];
    cout << "Enter elements : ";
    for (int i = 0; i < n; i++) {
        cin>>arr[i];
    }
    cout<<"Array before sort: ";
    print(arr,n);
    mergeSort(arr,0,n-1);
    cout<<"Array after sort: ";
    print(arr,n);    
    return 0;
}

File: Merge_Sort/MergeSort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MergeSort
{
    class Program
    {
        static int[] MergeSort(int[] arr)
        {
            int[] left;
            int[] right;
            int[] result = new int[arr.Length];

            if (arr.Length <= 1)
                return arr;

            int mid = arr.Length / 2;

            left = new int[mid];
            if (arr.Length % 2 == 0)
                right = new int[mid];
            else
                right = new int[mid + 1];

            for (int i = 0; i < left.Length; i++)
            {
                left[i] = arr[i];
            }

            for (int j = 0; j < right.Length; j++)
            {
                right[j] = arr[mid + j];
            }

            left = MergeSort(left);
            right = MergeSort(right);
            result = Merge(left, right);
            return result;
        }

        static int[] Merge(int[] left, int[] right)
        {
            int[] result = new int[left.Length + right.Length];
            int leftIndex = 0, rightIndex = 0, resultIndex = 0;

            //While either array still has an element
            while (leftIndex < left.Length || rightIndex < right.Length)
            {
                //If both arrays have elements
                if (leftIndex < left.Length && rightIndex < right.Length)
                {
                    if (left[leftIndex] <= right[rightIndex])
                    {
                        result[resultIndex] = left[leftIndex];
                        leftIndex++;
                        resultIndex++;
                    }
                    else
                    {
                        result[resultIndex] = right[rightIndex];
                        rightIndex++;
                        resultIndex++;
                    }
                }
                //If only the left array has elements
                else if (leftIndex < left.Length)
                {
                    result[resultIndex] = left[leftIndex];
                    leftIndex++;
                    resultIndex++;
                }
                // If only the right array has elements
                else if (rightIndex < right.Length)
                {
                    result[resultIndex] = right[rightIndex];
                    rightIndex++;
                    resultIndex++;
                }
            }
            return result;
        }

        static void Main(string[] args)
        {
            Console.WriteLine("Enter elements of the array");
            int[] arr = Array.ConvertAll(Console.ReadLine().Split(' '), int.Parse);
            int[] result = MergeSort(arr);
            Console.WriteLine("After merge sort, the array is");
            foreach (int item in result)
            {
                Console.Write(item + " ");
            }

            Console.Read();
        }
    }
}
